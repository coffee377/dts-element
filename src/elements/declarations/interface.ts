import {Container} from '../../collections';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_members} from '../../helpers/emit-members';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ObjectMember} from '../members/object';
import {GenericType} from '../types/generic';

export interface IInterfaceDeclarationRequiredParameters {
  name: string;
}

export interface IInterfaceDeclarationOptionalParameters {
  generics: GenericType[];
  members: ObjectMember[];
}

export class InterfaceDeclaration
    extends Declaration<IInterfaceDeclarationRequiredParameters, IInterfaceDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IInterfaceDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      members: [],
    });
  }

  public _emit_raw(container: Container): string {
    const {name, members, generics} = this.parameters;
    const generic = emit_generics(generics, this);
    return `interface ${name}${generic} ${emit_members(members, this)}`;
  }

}
