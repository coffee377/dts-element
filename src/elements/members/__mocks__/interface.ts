import {Member} from '../../member';

export class InterfaceMember extends Member<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[InterfaceMember ${this.parameters.owned.emit()}]`;
  }

}
