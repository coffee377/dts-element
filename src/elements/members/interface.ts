import {AnyElement} from '../../element';
import {emit_readonly} from '../../helpers/emit-readonly';
import {Member} from '../member';

// tslint:disable-next-line no-empty-interface
export interface IInterfaceMemberRequiredParameters {}

export interface IInterfaceMemberOptionalParameters {
  readonly: boolean;
}

export class InterfaceMember extends Member<IInterfaceMemberRequiredParameters, IInterfaceMemberOptionalParameters> {

  public get default_parameters(): IInterfaceMemberOptionalParameters {
    return {
      readonly: false,
    };
  }

  public _emit(_container: AnyElement | null): string {
    const {owned} = this.parameters;
    const readonly = emit_readonly(this.parameters.readonly, owned);
    return `${owned._emit_jsdoc()}${readonly}${owned._emit_raw(this)}`;
  }

}
