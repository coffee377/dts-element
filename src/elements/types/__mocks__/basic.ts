export class BasicType {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[BasicType ${this.parameters.name}]`;
  }

}
