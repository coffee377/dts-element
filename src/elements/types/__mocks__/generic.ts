export class GenericType {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[GenericType ${this.parameters.name}]`;
  }

}
