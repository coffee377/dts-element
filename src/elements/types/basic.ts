import {AnyElement} from '../../element';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_container: AnyElement | null): string {
    return this.parameters.name;
  }

}
