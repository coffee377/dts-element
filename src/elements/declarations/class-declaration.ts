import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_members} from '../../helpers/emit-members';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ClassMember} from '../members/class-member';
import {ClassType} from '../types/class-type';
import {GenericType} from '../types/generic-type';

export interface IClassDeclarationRequiredParameters {
  name: string;
}

export interface IClassDeclarationOptionalParameters {
  generics: GenericType[];
  members: ClassMember[];
  extends: ClassType | null;
  abstract: boolean;
}

export class ClassDeclaration
    extends Declaration<IClassDeclarationRequiredParameters, IClassDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IClassDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      members: [],
      extends: null,
      abstract: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {name, members, generics} = this.parameters;
    const generic = emit_generics(generics, this);
    const abstract = emit_abstract(this.parameters.abstract, this);
    const an_extends = emit_extends(this.parameters.extends, this);
    return `${abstract}class ${name}${generic}${an_extends} ${emit_members(members, this)}`;
  }

}
