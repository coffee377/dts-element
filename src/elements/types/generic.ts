import {Container} from '../../collections';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generic_default} from '../../helpers/emit-generic-default';
import {InterfaceDeclaration} from '../declarations/interface';
import {Type} from '../type';

export interface IGenericTypeRequiredParameters {
  name: string;
}

export interface IGenericTypeOptionalParameters {
  extends: Type | null;
  default: Type | null;
}

export class GenericType extends Type<IGenericTypeRequiredParameters, IGenericTypeOptionalParameters> {

  public get default_parameters(): IGenericTypeOptionalParameters {
    return {
      extends: null,
      default: null,
    };
  }

  public _emit(container: Container): string {
    const {name} = this.parameters;
    const generic_extends = emit_extends(this.parameters.extends, container);
    const generic_default = emit_generic_default(this.parameters.default, container);
    return `${name}${generic_extends}${generic_default}`;
  }

}
