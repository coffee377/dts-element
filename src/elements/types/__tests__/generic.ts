jest.unmock('../generic');

import {InterfaceDeclaration} from '../../declarations/interface';
import {BasicType} from '../basic';
import {GenericType} from '../generic';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new GenericType({
      name: 'T',
      extends: new BasicType({name: 'A'}),
      default: new BasicType({name: 'B'}),
    }).emit()).toMatchSnapshot();
  });
});

describe('#_emit()', () => {
  describe('InterfaceDeclaration', () => {
    const container = Object.create(InterfaceDeclaration.prototype);
    it('should return correctly with extends and default', () => {
      expect(new GenericType({
        name: 'T',
        extends: new BasicType({name: 'A'}),
        default: new BasicType({name: 'B'}),
      })._emit(container)).toMatchSnapshot();
    });
    it('should return correctly without extends and default', () => {
      expect(new GenericType({name: 'T'})._emit(container)).toMatchSnapshot();
    });
  });
});
