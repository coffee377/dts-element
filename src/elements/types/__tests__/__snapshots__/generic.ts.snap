// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#_emit() InterfaceDeclaration should return correctly with extends and default 1`] = `"T extends [BasicType A] = [BasicType B]"`;

exports[`#_emit() InterfaceDeclaration should return correctly without extends and default 1`] = `"T"`;

exports[`#emit() should return correctly 1`] = `"T"`;
